### PR TITLE
Image upload

### PR DESCRIPTION
**Pull request by quick007**

### CODE REVIEW

File: imagekit.ts
Comment (Bloxs) [inline imagekit.ts:1]: Probably should be deleted as this was our test file

File: utils/db/misc.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export const isUUID = (uuid: string) => {
+  if (uuid.length != 36) return false;
+  const uuidRegex =
+    /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/gi;
```
Comment (Bloxs) [inline utils/db/misc.ts:4]: Definitely not yoinked from https://melvingeorge.me/blog/check-if-string-valid-uuid-regex-javascript

File: utils/db/kv.types.ts [hunk 2]
```diff
@@ -10,7 +10,12 @@ export interface Event {
   name: string;
   supportEmail: string;
   description?: string;
-  bannerImage?: string;
+  banner: {
+    path?: string;
```
Comment (Bloxs) [inline utils/db/kv.types.ts:14]: IMO url makes more sense

File: utils/db/kv.types.ts [hunk 3]
```diff
@@ -10,7 +10,12 @@ export interface Event {
...
+    id?: string;
```
Comment (Bloxs) [inline utils/db/kv.types.ts:15]: shouldn't we only allow one banner per event thus not requiring this?

File: utils/db/kv.types.ts [hunk 4]
```diff
@@ -10,7 +10,12 @@ export interface Event {
...
+    fill: boolean;
+    uploading: boolean;
```
Comment (Bloxs) [inline utils/db/kv.types.ts:17]: Don't exactly get what this is for, either url is null/undefined if there's no image or it's a string if there is

File: routes/api/events/image/upload.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    const promises: [UploadResponse, void?] = [
+      await imageKit.upload({
+        file,
+        fileName: crypto.randomUUID(),
+      }),
+    ];
+
+    if (event.value.banner.id) {
+      promises.push(await imageKit.deleteFile(event.value.banner.id));
```
Comment (Bloxs) [inline routes/api/events/image/upload.ts:58]: Wouldn't it make more sense to move this above the upload one so there's no array?

File: routes/api/events/image/upload.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        {
+          status: 400,
+        },
+      );
+    }
+
+    await kv.set(["event", eventID], {
+      ...event.value,
+      banner: {
+        ...event.value.banner,
```
Comment (Bloxs) [inline routes/api/events/image/upload.ts:80]: Uploading is never unset

File: routes/events/(pages)/organizing/index.tsx [hunk 7]
```diff
@@ -2,6 +2,7 @@ import { Event, getUser, kv } from "@/utils/db/kv.ts";
 import { AppContext } from "$fresh/server.ts";
 import { renderToString } from "preact-render-to-string";
 import CTA from "@/components/buttons/cta.tsx";
+import { imageKit } from "@/routes/api/events/image/upload.ts";
```
Comment (Bloxs) [inline routes/events/(pages)/organizing/index.tsx]: Should prob be moved to `lib`

File: routes/api/events/image/upload.ts [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+// @deno-types="npm:@types/imagekit"
+import ImageKit, { UploadResponse } from "imagekit";
+import { Event, getUser, kv } from "@/utils/db/kv.ts";
+import { isUUID } from "@/utils/db/misc.ts";
+
+export const imageKit = new ImageKit({
+  publicKey: Deno.env.get("IMAGEKIT_PUBLIC_KEY")!,
+  privateKey: Deno.env.get("IMAGEKIT_PRIVATE_KEY")!,
+  urlEndpoint: Deno.env.get("IMAGEKIT_URL_ENDPOINT")!,
+});
```
Comment (Bloxs) [inline routes/api/events/image/upload.ts]: Should prob be moved to `lib`

File: routes/events/(pages)/organizing/index.tsx [hunk 9]
```diff
@@ -34,18 +35,46 @@ export default async function Homepage(req: Request, ctx: AppContext) {
...
+          return (
+            <div className="rounded-md overflow-hidden">
+              {e.banner.path ? (
+                (() => {
+                  const url = imageKit.url({
+                    path: e.banner.path,
+                    transformation: [
+                      {
+                        width: "400",
+                        quality: "85",
```
Comment (Bloxs) [inline routes/events/(pages)/organizing/index.tsx:53]: Didn't you say we're limited on transformations with imagekit

File: deno.json [hunk 10]
```diff
@@ -29,7 +29,7 @@
     "@/": "./",
     "$tabler/": "https://deno.land/x/tabler_icons_tsx@0.0.4/tsx/",
     "yup": "https://esm.sh/yup@1.2.0",
-    "imagekit": "https://esm.sh/imagekit@4.1.3"
+    "imagekit": "npm:imagekit@4.1.3"
```
Comment (Bloxs) [inline deno.json:32]: why weren't we doing this

File: components/pickers/image.tsx [hunk 11]
```diff
@@ -122,9 +149,10 @@ export default function ImagePicker({
             )}
           </div>
           <Toggle
+            disabled={uploading}
             name="Stretch to Fit"
             description="Stretch your image to fit the width of banner box. The way an image stretchs will vary based on device screen size. "
-            setEnabled={setFill}
+            setEnabled={setStretch as StateUpdater<boolean>}
```
Comment (Bloxs) [inline components/pickers/image.tsx:155]: if this is now called stretch why is it called filled in the db?

File: utils/db/misc.ts (hunk 1 above)
Comment (quick007) [reply utils/db/misc.ts:4]: 😳

File: utils/db/kv.types.ts (hunk 2 above)
Comment (quick007) [reply utils/db/kv.types.ts:14]: I disagree. we're not storing a url, we're storing a path like /image-name.jpg, then passing that into imagekit for url generation

File: routes/api/events/image/upload.ts (hunk 5 above)
Comment (quick007) [reply routes/api/events/image/upload.ts:58]: could, but why not execute them at the same time?

File: utils/db/kv.types.ts (hunk 4 above)
Comment (quick007) [reply utils/db/kv.types.ts:17]: this is if there's already an image and the user is replacing it. it prevents them from putting in another image while also keeping the old one intact for an (admittedly very small) amount longer

File: routes/api/events/image/upload.ts (hunk 6 above)
Comment (quick007) [reply routes/api/events/image/upload.ts:80]: uhhh proof?

File: routes/events/(pages)/organizing/index.tsx (hunk 9 above)
Comment (quick007) [reply routes/events/(pages)/organizing/index.tsx:53]: the reason we're using imagekit is bc it's (more or less) unlimited.

File: utils/db/kv.types.ts (hunk 2 above)
Comment (Bloxs) [reply utils/db/kv.types.ts:14]: > I disagree. we're not storing a url, we're storing a path like /image-name.jpg, then passing that into imagekit for url generation

Ah in that case then it's fine

File: routes/api/events/image/upload.ts (hunk 5 above)
Comment (Bloxs) [reply routes/api/events/image/upload.ts:58]: Shouldn't we delete the previous image first so imagekit doesn't throw any potential errors

File: routes/api/events/image/upload.ts (hunk 6 above)
Comment (Bloxs) [reply routes/api/events/image/upload.ts:80]: my proof is that I made it the fuck up

File: utils/db/kv.types.ts (hunk 3 above)
Comment (quick007) [reply utils/db/kv.types.ts:15]: you need the id to delete

File: routes/api/events/image/upload.ts (hunk 5 above)
Comment (quick007) [reply routes/api/events/image/upload.ts:58]: won't throw errors

File: components/pickers/image.tsx (hunk 11 above)
Comment (quick007) [reply components/pickers/image.tsx:155]: need multiple names

File: deno.json (hunk 10 above)
Comment (quick007) [reply deno.json:32]: idk you imported it

**Review by Bloxs — APPROVED** [review]
Hopefully it wont break anything

**Pull request merged**